File: Plugins/ScriptPlugins/VPNDetection.js
let vpnExceptionIds = [];
const vpnAllowListKey = 'Webfront::Nav::Admin::VPNAllowList';
const vpnWhitelistKey = 'Webfront::Profile::VPNWhitelist';

const init = (registerNotify, serviceResolver, configWrapper, pluginHelper) => {
    registerNotify('IManagementEventSubscriptions.ClientStateAuthorized', (authorizedEvent, token) => plugin.onClientAuthorized(authorizedEvent, token));

    plugin.onLoad(serviceResolver, configWrapper, pluginHelper);
    return plugin;
};

const plugin = {
    author: 'RaidMax',
    version: '2.1',
    name: 'VPN Detection Plugin',
    manager: null,
    configWrapper: null,
    logger: null,
    serviceResolver: null,
    translations: null,
    pluginHelper: null,
    enabled: true,

    commands: [{
        name: 'whitelistvpn',
        description: 'whitelists a player\'s client id from VPN detection',
        alias: 'wv',
        permission: 'SeniorAdmin',
        targetRequired: true,
        arguments: [{
            name: 'player',
            required: true
        }],
        execute: (gameEvent) => {
            vpnExceptionIds.push(gameEvent.Target.ClientId);
            plugin.configWrapper.setValue('vpnExceptionIds', vpnExceptionIds);

            gameEvent.origin.tell(`Successfully whitelisted ${gameEvent.target.name}`);
        }
    },
    {
        name: 'disallowvpn',
        description: 'disallows a player from connecting with a VPN',
        alias: 'dv',
        permission: 'SeniorAdmin',
        targetRequired: true,
        arguments: [{
            name: 'player',
            required: true
        }],
        execute: (gameEvent) => {
            vpnExceptionIds = vpnExceptionIds.filter(exception => parseInt(exception) !== parseInt(gameEvent.Target.ClientId));
            plugin.configWrapper.setValue('vpnExceptionIds', vpnExceptionIds);

            gameEvent.origin.tell(`Successfully disallowed ${gameEvent.target.name} from connecting with VPN`);
        }
    }
    ],

    interactions: [{
        // registers the profile action
        name: vpnWhitelistKey,
        action: function (targetId, game, token) {
            const helpers = importNamespace('SharedLibraryCore.Helpers');
            const interactionData = new helpers.InteractionData();

            interactionData.actionPath = 'DynamicAction';
            interactionData.interactionId = vpnWhitelistKey;
            interactionData.entityId = targetId;
            interactionData.minimumPermission = 3;
            interactionData.source = plugin.name;
            interactionData.actionMeta.add('InteractionId', 'command'); // indicate we're wanting to execute a command
            interactionData.actionMeta.add('ShouldRefresh', true.toString()); // indicates that the page should refresh after performing the action

            if (vpnExceptionIds.includes(targetId)) {
                interactionData.name = plugin.translations['WEBFRONT_VPN_BUTTON_DISALLOW']; // text for the profile button
                interactionData.displayMeta = 'ph-x-circle';

                interactionData.actionMeta.add('Data', `disallowvpn`); // command to execute
                interactionData.actionMeta.add('ActionButtonLabel', plugin.translations['WEBFRONT_VPN_ACTION_DISALLOW_CONFIRM']); // confirm button on the dialog
                interactionData.actionMeta.add('Name', plugin.translations['WEBFRONT_VPN_ACTION_DISALLOW_TITLE']); // title on the confirm dialog
            } else {
                interactionData.name = plugin.translations['WEBFRONT_VPN_ACTION_ALLOW']; // text for the profile button
                interactionData.displayMeta = 'ph-check-circle';

                interactionData.actionMeta.add('Data', `whitelistvpn`); // command to execute
                interactionData.actionMeta.add('ActionButtonLabel', plugin.translations['WEBFRONT_VPN_ACTION_ALLOW_CONFIRM']); // confirm button on the dialog
                interactionData.actionMeta.add('Name', plugin.translations['WEBFRONT_VPN_ACTION_ALLOW_TITLE']); // title on the confirm dialog
            }

            return interactionData;
        }
    },
    {
        name: vpnAllowListKey,
        action: function (targetId, game, token) {
            const helpers = importNamespace('SharedLibraryCore.Helpers');
            const interactionData = new helpers.InteractionData();

            interactionData.name = plugin.translations['WEBFRONT_NAV_VPN_TITLE']; // navigation link name
            interactionData.description = plugin.translations['WEBFRONT_NAV_VPN_DESC']; // alt and title
            interactionData.displayMeta = 'ph-check-circle'; // nav icon
            interactionData.interactionId = vpnAllowListKey;
            interactionData.minimumPermission = 3; // moderator
            interactionData.interactionType = 2; // 1 is RawContent for apis etc..., 2 is 
            interactionData.source = plugin.name;

            interactionData.scriptAction = (sourceId, targetId, game, meta, token) => {
                const clientsData = plugin.getClientsData(vpnExceptionIds);

                let table = '<table class="w-full text-left border-collapse">';

                const disallowInteraction = {
                    InteractionId: 'command',
                    Data: 'disallowvpn',
                    ActionButtonLabel: plugin.translations['WEBFRONT_VPN_ACTION_DISALLOW_CONFIRM'],
                    Name: plugin.translations['WEBFRONT_VPN_ACTION_DISALLOW_TITLE']
                };

                if (clientsData.length === 0) {
                    table += `<tr><td colspan="2" class="px-6 py-8 text-center text-muted">No players are whitelisted.</td></tr>`;
                }

                clientsData.forEach(client => {
                    table += `<tr class="border-t border-line hover:bg-surface-hover/30 transition-colors">
                                    <td class="px-6 py-4 whitespace-nowrap">
                                        <a href="/Client/Profile/${client.clientId}" class="text-sm font-medium hover:text-primary transition-colors">${client.currentAlias.name.stripColors()}</a>
                                    </td>
                                    <td class="px-6 py-4 text-right">
                                        <button type="button" class="profile-action cursor-pointer" data-action="DynamicAction" data-action-id="${client.clientId}"
                                           data-action-meta="${encodeURI(JSON.stringify(disallowInteraction))}">
                                            <div class="inline-flex items-center px-3 py-1.5 rounded-lg bg-red-600/20 text-red-400 border border-red-500/30 hover:bg-red-600/30 transition-colors text-sm font-medium">
                                                <i class="ph ph-x-circle mr-2 text-sm"></i>
                                                <span class="truncate">${plugin.translations['WEBFRONT_VPN_BUTTON_DISALLOW']}</span>
                                            </div>
                                        </button>
                                    </td>
                                </tr>`;
                });

                table += '</table>';

                return table;
            };

            return interactionData;
        }
    }
    ],

    onClientAuthorized: async function (authorizeEvent, token) {
        if (authorizeEvent.client.isBot || !this.enabled) {
            return;
        }
        await this.checkForVpn(authorizeEvent.client, token);
    },

    onLoad: function (serviceResolver, configWrapper, pluginHelper) {
        this.serviceResolver = serviceResolver;
        this.configWrapper = configWrapper;
        this.pluginHelper = pluginHelper;
        this.manager = this.serviceResolver.resolveService('IManager');
        this.logger = this.serviceResolver.resolveService('ILogger', ['ScriptPluginV2']);
        this.translations = this.serviceResolver.resolveService('ITranslationLookup');

        this.configWrapper.setName(this.name); // use legacy key
        this.configWrapper.getValue('vpnExceptionIds').forEach(element => vpnExceptionIds.push(parseInt(element)));
        this.logger.logInformation(`Loaded ${vpnExceptionIds.length} ids into whitelist`);

        this.enabled = this.configWrapper.getValue('enabled', newValue => {
            if (newValue) {
                plugin.logger.logInformation('{Name} configuration updated. Enabled={Enabled}', newValue);
                plugin.enabled = newValue;
            }
        });

        if (this.enabled === undefined) {
            this.configWrapper.setValue('enabled', true);
            this.enabled = true;
        }

        this.interactionRegistration = this.serviceResolver.resolveService('IInteractionRegistration');
        this.interactionRegistration.unregisterInteraction(vpnWhitelistKey);
        this.interactionRegistration.unregisterInteraction(vpnAllowListKey);

        this.logger.logInformation('{Name} {Version} by {Author} loaded. Enabled={Enabled}', this.name, this.version,
            this.author, this.enabled);
    },

    checkForVpn: async function (origin, _) {
        let exempt = false;
        // prevent players that are exempt from being kicked
        vpnExceptionIds.forEach(function (id) {
            if (parseInt(id) === parseInt(origin.clientId)) {
                exempt = true;
                return false;
            }
        });

        if (exempt) {
            this.logger.logInformation(`{origin} is whitelisted, so we are not checking VPN status`, origin);
            return;
        }

        if (origin.IPAddressString === null) {
            this.logger.logDebug('{Client} does not have an IP Address yet, so we are no checking their VPN status', origin);
        }

        const userAgent = `IW4MAdmin-${this.manager.getApplicationSettings().configuration().id}`;
        const stringDict = System.Collections.Generic.Dictionary(System.String, System.String);
        const headers = new stringDict();
        headers.add('User-Agent', userAgent);
        const pluginScript = importNamespace('IW4MAdmin.Application.Plugin.Script');
        const request = new pluginScript.ScriptPluginWebRequest(`https://api.xdefcon.com/proxy/check/?ip=${origin.IPAddressString}`,
            null, 'GET', 'application/json', headers);

        try {
            this.pluginHelper.requestUrl(request, (response) => this.onVpnResponse(response, origin));

        } catch (ex) {
            this.logger.logWarning('There was a problem checking client IP ({IP}) for VPN - {message}',
                origin.IPAddressString, ex.message);
        }
    },

    onVpnResponse: function (response, origin) {
        let parsedJSON = null;

        try {
            parsedJSON = JSON.parse(response);
        } catch {
            this.logger.logWarning('There was a problem checking client IP ({IP}) for VPN - {message}',
                origin.IPAddressString, response);
            return;
        }

        const usingVPN = parsedJSON.success && parsedJSON.proxy;

        if (usingVPN) {
            this.logger.logInformation('{origin} is using a VPN ({ip})', origin.toString(), origin.IPAddressString);
            const contactUrl = this.manager.getApplicationSettings().configuration().contactUri;
            let additionalInfo = '';
            if (contactUrl) {
                additionalInfo = this.translations['SERVER_KICK_VPNS_NOTALLOWED_INFO'] + ' ' + contactUrl;
            }
            origin.kick(this.translations['SERVER_KICK_VPNS_NOTALLOWED'] + ' ' + additionalInfo, origin.currentServer.asConsoleClient());
        } else {
            this.logger.logDebug('{Client} is not using a VPN', origin);
        }
    },

    getClientsData: function (clientIds) {
        const contextFactory = this.serviceResolver.resolveService('IDatabaseContextFactory');
        const context = contextFactory.createContext(false);
        const clientSet = context.clients;
        const clients = clientSet.getClientsBasicData(clientIds);
        context.dispose();

        return clients;
    }
};

File: Plugins/ScriptPlugins/VPNDetection.cs
#:package RaidMax.IW4MAdmin.SharedLibraryCore@2026.1.6.1

using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Http;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;
using Data.Abstractions;
using Microsoft.EntityFrameworkCore;
using SharedLibraryCore.Database.Models;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Logging;
using SharedLibraryCore;
using SharedLibraryCore.Commands;
using SharedLibraryCore.Configuration;
using SharedLibraryCore.Events.Management;
using SharedLibraryCore.Helpers;
using SharedLibraryCore.Interfaces;
using SharedLibraryCore.Interfaces.Events;

/// <summary>
/// VPN Detection Plugin - checks connecting clients against a proxy/VPN lookup service and kicks
/// anyone using a VPN unless they have been whitelisted. Ported from the legacy VPNDetection.js
/// Jint script with a C#-first design (typed config, HttpClient, command classes, inlined EF query).
/// </summary>
public class VpnDetectionPlugin : IPluginV2
{
    public static void RegisterDependencies(IServiceCollection serviceCollection)
    {
        serviceCollection.AddConfiguration("VPNDetectionSettings", new VpnDetectionConfiguration());
    }

    public string Name => "VPN Detection Plugin";
    public string Author => "RaidMax";
    public string Version => "2.1";

    private const string VpnWhitelistKey = "Webfront::Profile::VPNWhitelist";
    private const string VpnAllowListKey = "Webfront::Nav::Admin::VPNAllowList";

    private static readonly HttpClient HttpClient = new() { Timeout = TimeSpan.FromSeconds(10) };

    private readonly ILogger<VpnDetectionPlugin> _logger;
    private readonly VpnDetectionConfiguration _config;
    private readonly IInteractionRegistration _interactionRegistration;
    private readonly IDatabaseContextFactory _contextFactory;
    private readonly ApplicationConfiguration _appConfig;

    public VpnDetectionPlugin(
        ILogger<VpnDetectionPlugin> logger,
        VpnDetectionConfiguration config,
        IInteractionRegistration interactionRegistration,
        IDatabaseContextFactory contextFactory,
        IManager manager)
    {
        _logger = logger;
        _config = config;
        _interactionRegistration = interactionRegistration;
        _contextFactory = contextFactory;
        _appConfig = manager.GetApplicationSettings().Configuration();

        IManagementEventSubscriptions.ClientStateAuthorized += OnClientAuthorized;

        RegisterInteractions();

        _logger.LogInformation("{Name} {Version} by {Author} loaded. Enabled={Enabled}, Whitelisted={Count}",
            Name, Version, Author, _config.Enabled, _config.VpnExceptionIds.Count);
    }

    private void RegisterInteractions()
    {
        _interactionRegistration.UnregisterInteraction(VpnWhitelistKey);
        _interactionRegistration.UnregisterInteraction(VpnAllowListKey);

        // Per-profile toggle button: whitelist a client, or disallow an already-whitelisted one.
        _interactionRegistration.RegisterInteraction(VpnWhitelistKey, (clientId, game, token) =>
        {
            var loc = Utilities.CurrentLocalization.LocalizationIndex;
            var interaction = new InteractionData
            {
                InteractionId = VpnWhitelistKey,
                ActionPath = "DynamicAction",
                EntityId = clientId,
                MinimumPermission = EFClient.Permission.Moderator,
                Source = Name,
                ActionMeta =
                {
                    ["InteractionId"] = "command",
                    ["ShouldRefresh"] = "true"
                }
            };

            if (clientId.HasValue && _config.VpnExceptionIds.Contains(clientId.Value))
            {
                interaction.Name = loc["WEBFRONT_VPN_BUTTON_DISALLOW"];
                interaction.DisplayMeta = "ph-x-circle";
                interaction.ActionMeta["Data"] = "disallowvpn";
                interaction.ActionMeta["ActionButtonLabel"] = loc["WEBFRONT_VPN_ACTION_DISALLOW_CONFIRM"];
                interaction.ActionMeta["Name"] = loc["WEBFRONT_VPN_ACTION_DISALLOW_TITLE"];
            }
            else
            {
                interaction.Name = loc["WEBFRONT_VPN_ACTION_ALLOW"];
                interaction.DisplayMeta = "ph-check-circle";
                interaction.ActionMeta["Data"] = "whitelistvpn";
                interaction.ActionMeta["ActionButtonLabel"] = loc["WEBFRONT_VPN_ACTION_ALLOW_CONFIRM"];
                interaction.ActionMeta["Name"] = loc["WEBFRONT_VPN_ACTION_ALLOW_TITLE"];
            }

            return Task.FromResult<IInteractionData>(interaction);
        });

        // Admin nav page listing whitelisted clients, each with a disallow button.
        _interactionRegistration.RegisterInteraction(VpnAllowListKey, (clientId, game, token) =>
        {
            var loc = Utilities.CurrentLocalization.LocalizationIndex;
            var interaction = new InteractionData
            {
                Name = loc["WEBFRONT_NAV_VPN_TITLE"],
                Description = loc["WEBFRONT_NAV_VPN_DESC"],
                DisplayMeta = "ph-check-circle",
                InteractionId = VpnAllowListKey,
                MinimumPermission = EFClient.Permission.Moderator,
                InteractionType = InteractionType.TemplateContent,
                Source = Name,
                Action = async (sourceId, targetId, g, meta, ct) =>
                {
                    var clients = await GetClientsDataAsync(_config.VpnExceptionIds, ct);

                    var disallowInteraction = new Dictionary<string, string>
                    {
                        ["InteractionId"] = "command",
                        ["Data"] = "disallowvpn",
                        ["ActionButtonLabel"] = loc["WEBFRONT_VPN_ACTION_DISALLOW_CONFIRM"],
                        ["Name"] = loc["WEBFRONT_VPN_ACTION_DISALLOW_TITLE"]
                    };
                    var encodedMeta = Uri.EscapeDataString(JsonSerializer.Serialize(disallowInteraction));

                    // Rendered inside the webfront's card container; follows WebfrontCore/REDESIGN-GUIDE.md.
                    var table = $@"<div class=""flex items-center gap-3 px-4 py-2.5 border-b border-line"">
                            <h2 class=""flex-1 text-[11px] font-semibold uppercase tracking-wider text-muted"">{loc["WEBFRONT_NAV_VPN_TITLE"]}</h2>
                            <span class=""font-mono text-[11px] px-1.5 py-1 rounded-full bg-surface-alt text-subtle tabular-nums"">{clients.Count}</span>
                        </div>";

                    if (clients.Count == 0)
                    {
                        table += @"<div class=""flex flex-col items-center justify-center gap-2 py-10 text-muted text-sm"">
                                <i class=""ph ph-shield-check text-3xl text-secondary""></i>
                                <span>No players are whitelisted.</span>
                            </div>";
                    }
                    else
                    {
                        table += @"<div class=""overflow-x-auto""><table class=""w-full text-sm"">
                            <thead><tr class=""text-[10px] font-semibold uppercase tracking-wider text-muted"">
                                <th class=""text-left px-4 py-2.5 border-b border-line"">Player</th>
                                <th class=""text-left px-4 py-2.5 border-b border-line"">Client ID</th>
                                <th class=""text-right px-4 py-2.5 border-b border-line""></th>
                            </tr></thead><tbody class=""divide-y divide-line"">";

                        foreach (var client in clients)
                        {
                            var cleanName = System.Net.WebUtility.HtmlEncode(client.Name.StripColors());
                            table += $@"<tr class=""hover:bg-surface-hover transition-colors"">
                                    <td class=""px-4 py-2.5 align-middle"">
                                        <a href=""/client/{client.ClientId}"" class=""font-semibold text-foreground hover:text-primary transition-colors"">{cleanName}</a>
                                    </td>
                                    <td class=""px-4 py-2.5 align-middle font-mono tabular-nums text-muted"">#{client.ClientId}</td>
                                    <td class=""px-4 py-2.5 align-middle text-right"">
                                        <button type=""button"" class=""profile-action inline-flex items-center gap-1.5 h-8 px-3 rounded-lg border border-error/40 bg-surface-alt text-xs font-semibold text-error hover:bg-error/10 transition-colors""
                                                data-action=""DynamicAction"" data-action-id=""{client.ClientId}"" data-action-meta=""{encodedMeta}"">
                                            <i class=""ph ph-x-circle text-base""></i>
                                            <span>{loc["WEBFRONT_VPN_BUTTON_DISALLOW"]}</span>
                                        </button>
                                    </td>
                                </tr>";
                        }

                        table += "</tbody></table></div>";
                    }

                    return table;
                }
            };

            return Task.FromResult<IInteractionData>(interaction);
        });
    }

    private Task OnClientAuthorized(ClientStateAuthorizeEvent clientEvent, CancellationToken token)
    {
        if (clientEvent.Client.IsBot || !_config.Enabled)
        {
            return Task.CompletedTask;
        }

        return CheckForVpnAsync(clientEvent.Client, token);
    }

    private async Task CheckForVpnAsync(EFClient origin, CancellationToken token)
    {
        if (_config.VpnExceptionIds.Contains(origin.ClientId))
        {
            _logger.LogInformation("{Origin} is whitelisted, so we are not checking VPN status", origin.ToString());
            return;
        }

        if (origin.IPAddressString is null)
        {
            _logger.LogDebug("{Client} does not have an IP address yet, so we are not checking their VPN status",
                origin.ToString());
            return;
        }

        try
        {
            using var request = new HttpRequestMessage(HttpMethod.Get,
                $"https://api.xdefcon.com/proxy/check/?ip={origin.IPAddressString}");
            request.Headers.UserAgent.ParseAdd($"IW4MAdmin-{_appConfig.Id}");

            using var response = await HttpClient.SendAsync(request, token);
            var body = await response.Content.ReadAsStringAsync(token);

            bool usingVpn;
            try
            {
                using var json = JsonDocument.Parse(body);
                var root = json.RootElement;
                usingVpn = root.TryGetProperty("success", out var success) && success.ValueKind == JsonValueKind.True
                           && root.TryGetProperty("proxy", out var proxy) && proxy.ValueKind == JsonValueKind.True;
            }
            catch
            {
                _logger.LogWarning("There was a problem checking client IP ({IP}) for VPN - {Message}",
                    origin.IPAddressString, body);
                return;
            }

            if (!usingVpn)
            {
                _logger.LogDebug("{Client} is not using a VPN", origin.ToString());
                return;
            }

            _logger.LogInformation("{Origin} is using a VPN ({IP})", origin.ToString(), origin.IPAddressString);

            var loc = Utilities.CurrentLocalization.LocalizationIndex;
            var additionalInfo = string.IsNullOrEmpty(_appConfig.ContactUri)
                ? string.Empty
                : loc["SERVER_KICK_VPNS_NOTALLOWED_INFO"] + " " + _appConfig.ContactUri;
            var message = (loc["SERVER_KICK_VPNS_NOTALLOWED"] + " " + additionalInfo).TrimEnd();

            origin.Kick(message, origin.CurrentServer.AsConsoleClient());
        }
        catch (Exception ex)
        {
            _logger.LogWarning(ex, "There was a problem checking client IP ({IP}) for VPN", origin.IPAddressString);
        }
    }

    private async Task<List<ClientWhitelistEntry>> GetClientsDataAsync(IReadOnlyCollection<int> clientIds,
        CancellationToken token)
    {
        if (clientIds.Count == 0)
        {
            return [];
        }

        var ids = clientIds.ToList();
        await using var context = _contextFactory.CreateContext(false);
        return await context.Clients
            .Where(client => ids.Contains(client.ClientId))
            .Select(client => new ClientWhitelistEntry(client.ClientId, client.CurrentAlias.Name))
            .ToListAsync(token);
    }

    public void Dispose()
    {
        IManagementEventSubscriptions.ClientStateAuthorized -= OnClientAuthorized;
        _interactionRegistration.UnregisterInteraction(VpnWhitelistKey);
        _interactionRegistration.UnregisterInteraction(VpnAllowListKey);
        _logger.LogInformation("{Name} unloaded", Name);
    }

    private sealed record ClientWhitelistEntry(int ClientId, string Name);
}

/// <summary>
/// Whitelists a player's client id from VPN detection.
/// Usage: !whitelistvpn &lt;player&gt;
/// </summary>
public class WhitelistVpnCommand : Command
{
    private readonly VpnDetectionConfiguration _config;
    private readonly IConfigurationHandlerV2<VpnDetectionConfiguration> _configHandler;

    public WhitelistVpnCommand(CommandConfiguration config, ITranslationLookup translationLookup,
        VpnDetectionConfiguration scriptConfig, IConfigurationHandlerV2<VpnDetectionConfiguration> configHandler)
        : base(config, translationLookup)
    {
        Name = "whitelistvpn";
        Description = "whitelists a player's client id from VPN detection";
        Alias = "wv";
        Permission = EFClient.Permission.SeniorAdmin;
        RequiresTarget = true;
        Arguments =
        [
            new CommandArgument { Name = "player", Required = true }
        ];
        _config = scriptConfig;
        _configHandler = configHandler;
    }

    public override async Task ExecuteAsync(GameEvent gameEvent)
    {
        var targetId = gameEvent.Target.ClientId;
        if (!_config.VpnExceptionIds.Contains(targetId))
        {
            _config.VpnExceptionIds.Add(targetId);
            await _configHandler.Set(_config);
        }

        gameEvent.Origin.Tell($"Successfully whitelisted {gameEvent.Target.Name}");
    }
}

/// <summary>
/// Disallows a player from connecting with a VPN (removes them from the whitelist).
/// Usage: !disallowvpn &lt;player&gt;
/// </summary>
public class DisallowVpnCommand : Command
{
    private readonly VpnDetectionConfiguration _config;
    private readonly IConfigurationHandlerV2<VpnDetectionConfiguration> _configHandler;

    public DisallowVpnCommand(CommandConfiguration config, ITranslationLookup translationLookup,
        VpnDetectionConfiguration scriptConfig, IConfigurationHandlerV2<VpnDetectionConfiguration> configHandler)
        : base(config, translationLookup)
    {
        Name = "disallowvpn";
        Description = "disallows a player from connecting with a VPN";
        Alias = "dv";
        Permission = EFClient.Permission.SeniorAdmin;
        RequiresTarget = true;
        Arguments =
        [
            new CommandArgument { Name = "player", Required = true }
        ];
        _config = scriptConfig;
        _configHandler = configHandler;
    }

    public override async Task ExecuteAsync(GameEvent gameEvent)
    {
        var targetId = gameEvent.Target.ClientId;
        if (_config.VpnExceptionIds.RemoveAll(id => id == targetId) > 0)
        {
            await _configHandler.Set(_config);
        }

        gameEvent.Origin.Tell($"Successfully disallowed {gameEvent.Target.Name} from connecting with VPN");
    }
}

/// <summary>
/// Configuration for the VPN Detection plugin.
/// </summary>
public class VpnDetectionConfiguration
{
    /// <summary>
    /// Indicates whether VPN checking is enabled.
    /// </summary>
    public bool Enabled { get; set; } = true;

    /// <summary>
    /// Client ids that are exempt from VPN detection.
    /// </summary>
    public List<int> VpnExceptionIds { get; set; } = new();
}
